src/wallet/rpcwallet.cpp

```diff
@@ -4475,12 +4479,26 @@ static RPCHelpMan upgradewallet()
         version = request.params[0].get_int();
     }
 
+    std::string result;
     bilingual_str error;
     std::vector<bilingual_str> warnings;
+    const int previous_version{pwallet->GetVersion()};
+
     if (!pwallet->UpgradeWallet(version, error, warnings)) {
         throw JSONRPCError(RPC_WALLET_ERROR, error.original);
+    } else if (version == previous_version) {
+        result = "wallet version unchanged";
+    } else {
+        result = strprintf("wallet upgraded successfully from version %i to version %i", previous_version, version);
     }
+
     UniValue obj(UniValue::VOBJ);
+    obj.pushKV("wallet_name", pwallet->GetName());
+    obj.pushKV("previous_version", previous_version);
+    obj.pushKV("current_version", version);
+    if (!result.empty()) {
+        obj.pushKV("result", result);
+    }
     if (!error.empty()) {
         obj.pushKV("error", error.original);
```

Comment:
It looks like this is dead code, and always has been.

It would be easier to have one way to deal with errors. Either return them as `JSONRPCError` or as `{"error":...}`, but not both. Also, the error approach should be documented.

Reply:
Yes, I plan to have a look at this later today. That's the second part of my review feedback from #20282.

In #20391 I proposed returning an error field rather than raising an RPC error for this type of errors but wasn't sure which is preferred.

Reply:
I think either is fine as long as it is documented and consistent. An RPCError makes sense when the return object would otherwise be empty and would have to omit a lot of optional fields, except for the error.

Reply:
Proposing this:
```
$ ./src/bitcoin-cli -signet -rpcwallet="blank" upgradewallet 5
{
  "wallet_name": "blank",
  "previous_version": 169900,
  "current_version": 169900,
  "error": "Cannot downgrade wallet from version 169900 to version 5"
}
```

Reply:
(currently that returns):
```
error code: -4
error message:
Cannot downgrade wallet
```

Reply:
Done

Reply:
Let me know if there is anything needed WRT documenting the error approach.